ecr/object.ecr

```diff
@@ -20,127 +13,12 @@ module <%= namespace_name %>
-            {% resolved_type = arg.type.resolve %}
-            {% if arg.type.resolve == String %}
-              {{ arg.var }} = String.new({{ arg.var }})
-            {% elsif arg.type.resolve == Bool %}
-              {{ arg.var }} = {{ arg.var }} != 0
-            {% end %}
-            {% end %}
-
-            ::Box(LeanProc).unbox(_lib_box).call({{ signature.args.map(&.var).splat }})
-          }.pointer
-          {% end %}
-
-          handler = LibGObject.g_signal_connect_data(@source, name, handler,
-            GICrystal::ClosureDataManager.register(_box), ->GICrystal::ClosureDataManager.deregister, after.to_unsafe)
-          GObject::SignalConnection.new(@source, handler)
-        end
-
-        def emit({{ signature.args.splat }})
-          LibGObject.g_signal_emit_by_name(@source, {{ signature.name.stringify }}, {{ signature.args.map(&.var).splat }})
-        end
-      end
-
-      def {{ signature.name }}_signal
-        {{ signature.name.titleize }}Signal.new(self)
-      end
-
-      def self._class_init(klass : Pointer(LibGObject::TypeClass), user_data : Pointer(Void)) : Nil
-        LibGObject.g_signal_new({{ signature.name.stringify }}, g_type,
-        GObject::SignalFlags.flags(RunLast, NoRecurse, NoHooks),
-        0,                   # class_offset
-        Pointer(Void).null,  # accumulator
-        Pointer(Void).null,  # accumulator_data
-        Pointer(Void).null,  # marshaller
-        GObject::TYPE_NONE,  # return_type
-        {{ signature.args.size }}, # n_params
-        {% for arg in signature.args %}
-        {{ arg.type }}.g_type,
-        {% end %}
-        Pointer(Void).null)
-        previous_def
-      end
-    end
-  <% end %>
-
     <% if object.class_struct.nil? || g_object_type? # This is just to avoid render a empty macro block %>
     macro inherited
     <% if object.class_struct.nil? %>
     {{ raise "Cannot inherit from #{@type.superclass}" unless @type.annotation(GObject::GeneratedWrapper) }}
     <% elsif g_object_type? %>
       {% unless @type.annotation(GObject::GeneratedWrapper) %}
-      # GType for the new created type
-      @@_g_type : UInt64 = 0
-
-      def self.g_type : UInt64
-        if LibGLib.g_once_init_enter(pointerof(@@_g_type)) != 0
-          g_type = {{ @type.superclass.id }}._register_derived_type({{ @type.id }},
-            ->_class_init(Pointer(LibGObject::TypeClass), Pointer(Void)),
-            ->_instance_init(Pointer(LibGObject::TypeInstance), Pointer(LibGObject::TypeClass)))
-
-          LibGLib.g_once_init_leave(pointerof(@@_g_type), g_type)
-        end
-
-        @@_g_type
-      end
-
-      # :nodoc:
-      def self._class_init(klass : Pointer(LibGObject::TypeClass), user_data : Pointer(Void)) : Nil
-      end
-
-      # :nodoc:
-      def self._instance_init(instance : Pointer(LibGObject::TypeInstance), type : Pointer(LibGObject::TypeClass)) : Nil
-      end
-
       # Cast a `GObject::Object` to this type, returns nil if cast can't be made.
```

Comment:
I would also really like to get rid of this special case, but it uses `<%= object.qdata_get_func %>`, which I have no access to in `object.cr`.

Maybe it would be better to add a `qdata_get_func` macro on all fundamental types or something like that? This would make the source cleaner.

Reply:
I'm talking about these lines:
https://github.com/hugopl/gi-crystal/blob/ce61e6ccdb08fd5ee36b7eecf0c71655a305fbe1/ecr/object.ecr#L144-L155

Reply:
We could add functions on GICrystal namespace:

`GICrystal.instance_pointer` with overloads for Object and ParamSpec and a similar approach for the other things we store in qdata.

But I guess they need to be in their own include file since when gi_crystal/util.cr is included the Object and ParamSpec classes aren’t declared yet.

Reply:
I'm creating a patch for this...